internal/error.go

```diff
@@ -27,82 +27,98 @@ package internal
-			err.Details(&detailMsg) // extract strong typed details (corresponding to CustomErrReasonA)
-			// handle CustomErrReasonA
-		case CustomErrReasonB:
-			// handle CustomErrReasonB
+	var applicationErr *ApplicationError
+	if errors.As(err, &applicationError) {
+		// handle activity errors (created via NewApplicationError() API)
+		if !applicationErr.NonRetryable() {
+			// manually retry activity
+		}
+		var detailMsg string // assuming activity return error by NewApplicationError("message", true, "string details")
+		applicationErr.Details(&detailMsg) // extract strong typed details
+
+		// handle activity errors (errors created other than using NewApplicationError() API)
+		switch err.OriginalType() {
+		case "CustomErrTypeA":
+			// handle CustomErrTypeA
+		case CustomErrTypeB:
+			// handle CustomErrTypeB
 		default:
 			// newer version of activity could return new errors that workflow was not aware of.
 		}
-	case *workflow.GenericError:
-		// handle generic error (errors created other than using NewCustomError() API)
-	case *workflow.CanceledError:
+	}
+
+	var canceledErr *CanceledError
+	if errors.As(err, &canceledErr) {
 		// handle cancellation
-	case *workflow.TimeoutError:
-		// handle timeout, could check timeout type by err.TimeoutType()
-	case *workflow.PanicError:
-		// handle panic
+	}
+
+	var timeoutErr *TimeoutError
+	if errors.As(err, &timeoutErr) {
+		// handle timeout, could check timeout type by timeoutErr.TimeoutType()
+	}
+
+	var panicErr *PanicError
+	if errors.As(err, &panicErr) {
+		// handle panic, message and stack trace are available by panicErr.Error() and panicErr.StackTrace()
 	}
 }
 
-Errors from child workflow should be handled in a similar way, except that there should be no *PanicError from child workflow.
-When panic happen in workflow implementation code, temporal client library catches that panic and causing the decision timeout.
+Errors from child workflow should be handled in a similar way, except that instance of *ChildWorkflowExecutionError is returned to
+workflow code. It will contains *ActivityTaskError, which in turn will contains on of the errors above.
+When panic happen in workflow implementation code, SDK catches that panic and causing the decision timeout.
 That decision task will be retried at a later time (with exponential backoff retry intervals).
+
+Workflow consumers will get an instance of *WorkflowExecutionError. This error will contains one of errors above.
 */
 
 type (
-	// CustomError returned from workflow and activity implementations with reason and optional details.
-	CustomError struct {
-		reason  string
-		details Values
-	}
-
-	// GenericError returned from workflow/workflow when the implementations return errors other than from NewCustomError() API.
-	GenericError struct {
-		err string
+	// ApplicationError returned from activity implementations with message and optional details.
+	ApplicationError struct {
```

Comment:
What about deserialization errors?  Example unable to deserialize signal data.

Reply:
Deserialization of signal data is in user workflow/activity code, right? In case of any user error in workflow or activity code `ApplicationError` will be returned. Users can return:
1. `ApplicationError` with custom details and it will go through as is,
2. `MyApplicationError` (or any other error) which will be converted to `ApplicationError` and `"MyApplicationError"` string will be stored in `originalType` field,
3. `errors.New("my error")` which will be converted to `ApplicationError` and message will be stored in `message` field.

I think this should be enough to pass information to error handlers.

Reply:
I don't think user will have any control over returning this error.  This needs to be handled by SDK and returned back to application.  Here are some scenarios:

1. Deserialization of Activity parameters
2. Serialization of Activity completion result or failures
3. Serialization of Heartbeat data
4. Deserialization of Workflow input parameters
5. Serialization of Workflow completion and failures
6. Deserialization of signal data.  Needs to be communicated through signal channel returning back error.

Reply:
I think we should make is super straightforward to handle serialization/deserialization errors.

Reply:
Yes, you right. This is SDK error. Not user one. In this case it still will be converted to `ApplicationError` because all `errors.New/fmt.Errorf` are `ApplicationError`s.

The main idea behind this PR though is to convert errors to protobuf messages every time when they need to be transmitted between SDK and server. From this prospective all these errors are `ApplicationError`s. I can create special types for serialization/deserialization errors to simplify error handling in workflow code but let's discuss it first and this gonna be separate PR.